### PR TITLE
Add check blocking PRs to protected files (e.g. package.json)

### LINKED ISSUE #25144: Fail if PR changes root package.json or package-lock.json inadvertently
This PR inadvertently made changes to the root `package.json` and `package-lock.json`:

https://github.com/Azure/azure-rest-api-specs/pull/21994/commits/cdaebb9cf36f32204b68fe4b0561bdd86a53212a

Another example:

https://github.com/Azure/azure-rest-api-specs-pr/commit/277be0c84c3d41d8f1ab8b97c34fd77452ae2eb8

Ideally, this would fail.  However, there are valid scenarios to update `package.json` (like when upgrading TypeSpec), which should not fail.  So, it may not be possible to differentiate valid from invalid updates.

### PR DESCRIPTION
**Pull request by mikeharder**
- Fixes #25144

Example showing check failure: https://github.com/Azure/azure-rest-api-specs/pull/30350

### CODE REVIEW

File: .github/workflows/package-json-protection.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  package-json-protection:
+    name: package.json Protection
+
+    runs-on: ubuntu-latest
+
+    steps:
+    - name: Fail
+      uses: actions/github-script@v7
+      with:
+        script: core.setFailed("'package.json' and 'package-lock.json' should only be updated by the Azure SDK team.  If the changes are intentional, the PR may be merged via bypassing the branch protections.")
```
Comment (weshaggard) [inline .github/workflows/package-json-protection.yaml:21]:
```suggestion
        script: core.setFailed("'package.json' and 'package-lock.json' should only be updated by the Azure SDK team.  If the changes are intentional, the PR may be merged by the Azure SDK team via bypassing the branch protections.")
```

**Comment by weshaggard** [comment]
I suspect you are correct that if we make this required it will need to run on all PRs. 

Do you know how this will end up looking in the "Next Steps to merge"?

Another potential option is to not have this pipeline fail but instead add the same comment we would on the PR saying to undo this change unless you are a member of the Azure SDK team. It wouldn't block folks from merging but it would automate what we would have done.

**Comment by mikeharder** [comment]
> I suspect you are correct that if we make this required it will need to run on all PRs.

I will update the check to run on all PRs, use `ChangedFiles-Functions.ps1` to detect changes locally, and pass/fail based on the result.

> Do you know how this will end up looking in the "Next Steps to merge"?

Good question, I will test it.  I believe "Next Steps to merge" already has code to handle any required checks by default.

> Another potential option is to not have this pipeline fail but instead add the same comment we would on the PR saying to undo this change unless you are a member of the Azure SDK team. It wouldn't block folks from merging but it would automate what we would have done.

I think it's simpler (and more effective) to just fail the check.  We **really** don't want unintentional changes to these files.

File: .github/workflows/package-json-protection.yaml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+      with:
+        # Required since "HEAD^" is passed to Get-ChangedFiles
+        fetch-depth: 2
+
+    - name: Detect changes to package[-lock].json
+      run: |
+        . eng/scripts/ChangedFiles-Functions.ps1
+        $changedFiles = @(Get-ChangedFiles -baseCommitish HEAD^ -targetCommitish HEAD -diffFilter "")
+        if (($changedFiles -contains "package.json") -or ($changedFiles -contains "package-lock.json")) {
+          Write-Output "::error::'package.json' and 'package-lock.json' should only be updated by the Azure SDK team.  If the changes are intentional, the PR may be merged by the Azure SDK team via bypassing the branch protections."
```
Comment (weshaggard) [inline .github/workflows/package-json-protection.yaml:25]: https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#setting-a-notice-message

Should we create an annotation on the file so it shows up in the code review flow?

**Review by weshaggard — APPROVED** [review]

File: .github/workflows/package-json-protection.yaml (hunk 2 above)
Comment (mikeharder) [reply .github/workflows/package-json-protection.yaml]: Should be working now.  See the test PR to verify: https://github.com/Azure/azure-rest-api-specs/pull/30350

**Review by weshaggard — APPROVED** [review]

**Pull request merged**